ArduPlane/soaring.cpp

```diff
@@ -47,175 +47,134 @@ void Plane::update_soaring() {
+            g2.soaring_controller.update_cruising();
 
-    case Mode::Number::LOITER: {
-        // Update thermal estimate and check for switch back to AUTO
-        g2.soaring_controller.update_thermalling();  // Update estimate
+            if (g2.soaring_controller.check_thermal_criteria()) {
+                gcs().send_text(MAV_SEVERITY_INFO, "Soaring: Thermal detected, entering %s", mode_loiter.name());
 
-        Vector2f prev_wp, next_wp;
+                set_mode(mode_loiter, ModeReason::SOARING_THERMAL_DETECTED);
+            }
+            break;
 
-        // If previous mode was AUTO and there was a previous NAV command, we can use previous and next wps for drift calculation.
-        if (previous_mode->mode_number() == Mode::Number::AUTO) {
-            AP_Mission::Mission_Command current_nav_cmd = mission.get_current_nav_cmd();
-            AP_Mission::Mission_Command prev_nav_cmd;
+        case Mode::Number::LOITER: {
+            // Update thermal estimate and check for switch back to AUTO
+            g2.soaring_controller.update_thermalling();  // Update estimate
+
+            // Check distance to home.
+            Vector3f position;
+            if (!ahrs.get_relative_position_NED_home(position)) {
+                return;
+            } else if (g2.soaring_controller.max_radius >= 0 &&
+                       powf(position.x,2)+powf(position.y,2) > powf(g2.soaring_controller.max_radius,2) &&
+                    previous_mode->mode_number()!=Mode::Number::AUTO) {
+                // Some other loiter status, and outside of maximum soaring radius, and previous mode wasn't AUTO
+                gcs().send_text(MAV_SEVERITY_INFO, "Soaring: Outside SOAR_MAX_RADIUS, RTL");
+                set_mode(mode_rtl, ModeReason::SOARING_DRIFT_EXCEEDED);
+            }
 
-            if (!(mission.get_next_nav_cmd(mission.get_prev_nav_cmd_with_wp_index(), prev_nav_cmd) &&
-                   prev_nav_cmd.content.location.get_vector_xy_from_origin_NE(prev_wp) &&
-                current_nav_cmd.content.location.get_vector_xy_from_origin_NE(next_wp))) {
-                prev_wp.x = 0.0;
-                prev_wp.y = 0.0;
-                next_wp.x = 0.0;
-                next_wp.y = 0.0;
+            // If previous mode was AUTO and there was a previous NAV command, we can use previous and next wps for drift calculation.
+            Vector2f prev_wp, next_wp;
+
+            if (previous_mode->mode_number() == Mode::Number::AUTO) {
+                AP_Mission::Mission_Command current_nav_cmd = mission.get_current_nav_cmd();
+                AP_Mission::Mission_Command prev_nav_cmd;
+
+                if (!(mission.get_next_nav_cmd(mission.get_prev_nav_cmd_with_wp_index(), prev_nav_cmd) &&
+                       prev_nav_cmd.content.location.get_vector_xy_from_origin_NE(prev_wp) &&
+                    current_nav_cmd.content.location.get_vector_xy_from_origin_NE(next_wp))) {
+                    prev_wp.x = 0.0;
```

Comment:
We have `.zero()` methods on vectors.

Reply:
A short comment indicating why it's OK for prev_wp and next_wp to be zero would be good.  To my eye cruise critiera will always be false in that case.

Reply:
Switched to using .zero() and added comment.